packages/playwright-ct-react/registerSource.mjs

```diff
@@ -23,22 +23,48 @@ import { createRoot as __pwCreateRoot } from 'react-dom/client';
 /** @typedef {import('../playwright-test/types/experimentalComponent').Component} Component */
 /** @typedef {import('react').FunctionComponent} FrameworkComponent */
 
+/** @type {Map<string, () => Promise<FrameworkComponent>>} */
+const __pwLoaderRegistry = new Map();
+
 /** @type {Map<string, FrameworkComponent>} */
 const __pwRegistry = new Map();
+
 /** @type {Map<Element, import('react-dom/client').Root>} */
 const __pwRootRegistry = new Map();
 
 /**
- * @param {{[key: string]: FrameworkComponent}} components
+ * @param {{[key: string]: () => Promise<FrameworkComponent>}} components
  */
 export function pwRegister(components) {
   for (const [name, value] of Object.entries(components))
-    __pwRegistry.set(name, value);
+    __pwLoaderRegistry.set(name, value);
 }
 
 /**
  * @param {Component} component
  */
+async function __pwResolveComponent(component) {
+  if (typeof component !== 'object' || Array.isArray(component))
+    return
+
+  let componentFuncLoader = __pwLoaderRegistry.get(component.type);
+  if (!componentFuncLoader) {
+    // Lookup by shorthand.
+    for (const [name, value] of __pwLoaderRegistry) {
```

Comment:
Now that this code is here, do you still need it in pwRender?

Reply:
Can this be fixed by moving this code to `@playwright/experimental-ct-core` later on?

Reply:
Sure, we can follow up. But I don't think a lot of this code will end up in ct-core - there are more differences than similarities between the ports, generalizing them would be artificial.

Reply:
I'm not sure how this is related to the ct-core.

What I see here is that upon async `mount`, another async `__pwResolveComponent` is called, that is supposed to resolve the components. It makes sure that `__pwRegistry` is populated before `__pwRender` is called. So this seems like an ideal place to actually resolve the components, including those referenced by the shorthands that you process here already. I did not read too much into it, but doesn't it make sense?

Reply:
> that is supposed to resolve the components. It makes sure that __pwRegistry is populated before __pwRender is called.

Yes, correct. 

Maybe i'm missing something, but i thought of adding a couple of utils like a `onMount` function in `ct-core` which abstracts some things away like resolving the component beforehand and so on. See rough sketch:

Maybe I'm missing something, but i thought about adding some utilities like an `onMount` function in `ct-core` that abstracts some things like resolving the component and so on. See rough sketch:

```ts
// registerSource.mjs
import { onMount, onUnmount, onUpdate } from '@playwright/experimental-ct-core';
import { FunctionComponent as Component } from 'react';

onMount<Component>(async (Component, meta, rootElement) => {
  // error handling, registries and before/after mount are inside ct-core
  const App = __pwCreateRoot(rootElement);
  root.render(() => __pwRender(Component, meta));

  return App;
});

onUpdate<Component>(async (App, Component) => {
  // error handling and registries are inside ct-core
  App.render(__pwRender(Component));
});

onUnmount<Component>(async (App) => {
  // error handling and registries are inside ct-core
  App.unmount();
});
```